lib/internal/test_runner/test.js

```diff
@@ -518,6 +522,14 @@ class Test extends AsyncResource {
     if (name === 'before' || name === 'after') {
       hook.run = runOnce(hook.run);
     }
+    if (name === 'before' && this.startTime !== null) {
+      // Test has already started, run the hook immediately
+      PromisePrototypeThen(hook.run(this.getRunArgs()), () => {
+        if (hook.error != null) {
+          this.fail(hook.error);
+        }
+      });
+    }
```

Comment:
1. I don't think that createHook should run the hook as it's confusing
2. Why did you need to add it here and the one in the run is not enough?

Reply:
it is indeed confusing, thus the existance of this bug.
for example:
```js
test('test', async (t) => {
  t.before(() => console.log('before'));
});
```
at the point of calling `t.before` `test.run` is already called for the test, so it is too late to run its hook

Reply:
After looking at the code again I got it, thanks

should we also run for tests that finished? (if it's in set timeout for some reason)